pkg/controller/servicebinding/binder.go

```diff
@@ -342,17 +368,25 @@ func (b *binder) appendEnvVar(
 	return updatedEnvList
 }
 
-// appendEnvFrom based on secret name and list of EnvFromSource instances, making sure secret is
+// updateEnvFromList based on secret name and list of EnvFromSource instances, making sure secret is
 // part of the list or appended.
-func (b *binder) appendEnvFrom(envList []corev1.EnvFromSource, secret string) []corev1.EnvFromSource {
-	for _, env := range envList {
-		if env.SecretRef != nil && env.SecretRef.Name == secret {
-			b.logger.Debug("Directive 'envFrom' is already present!")
-			// secret name is already referenced
-			return envList
+func (b *binder) updateEnvFromList(envList []corev1.EnvFromSource, secret string) []corev1.EnvFromSource {
+	var indices []int
+	for k, env := range envList {
+		if env.SecretRef != nil {
+			if env.SecretRef.Name == secret {
+				b.logger.Debug("Directive 'envFrom' is already present!")
+				return envList
+			} else if strings.Contains(env.SecretRef.Name, b.sbr.GetName()) {
+				indices = append(indices, k)
+			}
 		}
 	}
 
+	for _, v := range indices {
+		envList = append(envList[:v], envList[v+1:]...)
```

Comment:
`envList` is being updated using a list of valid indices only for the first iteration, similarly to `volumes` somewhere in this changelist.

Assuming `envList` contains `a`, `b`, `c` and `d` (indices `0` to `3`) and `indices` contains `1` and `3`, after the first iteration (removing index `1` from the list) `envList` will contain `a`, `c` and `d`. The second iteration (to remove index `3` from the list) will fail because the list now only contains indices `0`, `1` and `2`.